### ISSUE TITLE
Trammel / Felucca foliage.

### ISSUE THREAD

**Issue by ZaneDubya** [ZaneDubya/UltimaXNA#115]
In UltimaXNA, trees are alive, there are flowers everywhere, and the grass grows regardless of what world you are in. But in Felucca, trees should be bare and the ground should be littered with bones and gravestones.

**Comment by Vodacek** [comment]
original client changes flowers, leaves etc based on season packet. Excerpt from packet guide:

```
Seasonal Information(3 bytes) 0xBC
       BYTE cmd
       BYTE id1
       BYTE id2

Note: Server message
Note: if id2 = 1, then this is a season change.
Note: if season change, then id1 = (0=spring, 1=summer, 2=fall, 3=winter, 4 = desolation)
```

This behaviour can be adjusted from RunUO, for more details see https://github.com/runuo/runuo/blob/master/Scripts/Misc/MapDefinitions.cs

We tried to investigate this, but without any chance, this feature must be hardcoded inside original client, we change on our "shard" seasons 9 times a year and the best way to do this is simply modify map and static by your own

**Comment by naggie** [comment]
So as a project, what's the best route? Parity with UO original or not?

**Comment by PatrikSamuelTauchim** [comment]
@Vodacek @naggie I wish to have season feature working as expected. In original client, I have never seen it working.

Would be really nice to have working seasons, something like:
- 0=spring - everything rendered normally as in datafiles - flowers, foliages and so on. (option, if you dont want to use seasons on your server at all)
- 1=summer  - some flowers replaced with "grass"?
- 2=fall - green foliages (leaves) are replaced with brown foliages or removed if brown is not available. green grass replaced with brown grass.
- 3=winter - foliages (leaves) and grass removed. terrain grass replaced with snow
- 4 = desolation) - like original client?

Season specifics can be read from external file in format like: [ID] [ID for summer] [ID for fall][ID for winter]

**Comment by ZaneDubya** [comment]
@naggie My number one goal is feature parity with the legacy client. To that end we should get seasons working to the extent they do in legacy. Once that is done, we can add new features like better support for seasons. I for one would like to see something like what @Eskymak proposed as a 'new feature' which could be enabled/disabled by the player.

@Vodacek Thank you for posting that packet! Now we just need to find - if it exists - the translation table inside the client that changes the itemIDs of the statics you mentioned. If it is hardcoded in the client, we should be able to find an array of short integers with the itemIds to translate in there somewhere...

**Comment by ZaneDubya** [comment]
I've spent the past five hours looking for such a table - and have been unable to find it. If it exists.

Here's an alternate idea: we include a new table - made from scratch - in the UltimaXNA client.

On a seasonal change, the client would go through every loaded static item and change the itemID if it matched one of the entries in the table.

We would need to figure out which items need to be translated, and what item IDs they should be translated to. This wouldn't be difficult, but it would be time-consuming if we wanted to match the legacy client exactly.

Then again, since these graphics are only window dressing, we needn't match the legacy client exactly. If a flower turns into a mushroom, but it's not the exact same mushroom as in legacy, I doubt people would notice or care - and if they do, they can take the time to fix the translation table themselves.

**Comment by fdsprod** [comment]
@ZaneDubya TexTerr.def I dont know the format, but im pretty sure thats what defines seasonal tile changes.  5 seasons 285 lines / 5 = 57 lines per season?

**Comment by PatrikSamuelTauchim** [comment]
@jeffboulanger I just openned that file in notepad and there is really some list, but im unable to find itemids of items, that are getting translated. Maybe this file is for something else or IDs are in different format, which doestn make much sense?

**Comment by ZaneDubya** [comment]
TexTerr.def might be used to [translate land tiles](http://www.runuo.com/community/threads/cannot-edit-landtile-in-uofiddler.530831/). The format proposed in the linked thread is:

OriginalTileID {ReplacementTileID} unknown

**Comment by fdsprod** [comment]
@Eskymak thats because it isnt item id's its tile ids
@ZaneDubya I found a lot of other posts on RunUO and other map making sites explaining how to edit this file for transisions with seasons but I will look again tonight at home to see if i can determine if they are correct.

**Comment by PatrikSamuelTauchim** [comment]
@jeffboulanger yeah, if I was googleing before posting I would know that. I found it after that :-)

This file could be really useful for replacing grass tiles with snow tiles in winter for example. It is more comfortable, then edit mapX.mul file and replace terrain texture. But, not sure if this could somehow be related to seasons. And we are still missing ItemIDs definitions.

Is there any way how to get ItemID from already translated items? Props/get itemid commands doesnt work on them. Could be useful in manual making of the list.

**Comment by ZaneDubya** [comment]
Yeah, interesting that [props doesn't work on them - it does work on other static items, doesn't it?

**Comment by PatrikSamuelTauchim** [comment]
@ZaneDubya  Yes, on other statics items and terrain tiles it works. Also works on non translated items but once item is already translated (tree to gravestone), it does not work on it.

**Comment by ZaneDubya** [comment]
@Eskymak Which makes sense, I suppose, because RunUO doesn't know anything about seasonal translations! When you send a [target static item command to RunUO, it sends the itemID for the item you clicked on - and if RunUO doesn't have a static item with that itemID in the clicked tile, it doesn't do anything with the message. Neat.

So here's how I'm going to do it: just a big table of itemID translations. I'll do a couple of them perfectly, and then I'll just fudge the rest of them. If someone wants to make it match legacy perfectly in the future, they're more than welcome to take the time to do that themselves.

Hopefully done by the end of today.

[Update] There are at most 550 itemIDs that need to be translated. Quite possibly far fewer. Just translating ~200 items would cover 99.9% of all the translatable items in the map.

**Comment by ZaneDubya** [comment]
Off to a good start!

@jeffboulanger The SeasonalTranslator is a plugin. Thanks for setting that up! ![ss png](https://cloud.githubusercontent.com/assets/7041719/7898718/65ac1662-06ce-11e5-9cdf-a121312bccb4.jpg)

**Comment by Vodacek** [comment]
XNA on Windows 10? nice!
 

---

**Comment by ZaneDubya** [comment]
I'm marking this as complete, even though the translation tables aren't done. If you want to help with those, check out #170!
